### PR TITLE
.Net: VectorStore: Adding a generic data model and related azure ai search mapper.

### PR DESCRIPTION
**Pull request by westey-m**
### Motivation and Context

For cases where a developer is defining a schema using configuration, it's useful to have a generic data model that can be used with any data source.  A separate definition can be supplied in this case, that defines the schema, and this can look different to the data model.

### Description

- Adding a generic data model for any vector store records
- Adding a mapping for Azure AI Search that uses the definition to map from the database record to the generic data model.

#8536 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        foreach (var property in this._vectorStoreRecordDefinition.Properties)
+        {
+            if (property is VectorStoreRecordKeyProperty keyProperty)
+            {
+                var storagePropertyName = keyProperty.StoragePropertyName ?? keyProperty.DataModelPropertyName;
+                storageJsonObject.Add(storagePropertyName, dataModel.Key);
+            }
+            else if (property is VectorStoreRecordDataProperty dataProperty)
+            {
+                if (dataModel.Data != null)
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs:49]: We tend to favour `is not null`

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    /// <inheritdoc />
+    public VectorStoreGenericDataModel<string> MapFromStorageToDataModel(JsonObject storageModel, StorageToDataModelMapperOptions options)
+    {
+        Verify.NotNull(storageModel);
+
+        // Create variables to store the response properties.
+        var dataProperties = new Dictionary<string, object?>();
+        var vectorProperties = new Dictionary<string, object?>();
+        string key = string.Empty;
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs:79]: Should this be null?

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                    vectorProperties.Add(vectorProperty.DataModelPropertyName, vector);
+                }
+                else
+                {
+                    vectorProperties.Add(vectorProperty.DataModelPropertyName, null);
+                }
+            }
+        }
+
+        return new VectorStoreGenericDataModel<string>(key) { Data = dataProperties, Vectors = vectorProperties };
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs:124]: Do we need to test for a to make sure at a minimum a key value was set?

File: dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchGenericDataModelMapperTests.cs [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+                ["NullableLongDataProp"] = 4L,
+                ["FloatDataProp"] = 5.0f,
+                ["NullableFloatDataProp"] = 6.0f,
+                ["DoubleDataProp"] = 7.0,
+                ["NullableDoubleDataProp"] = 8.0,
+                ["BoolDataProp"] = true,
+                ["NullableBoolDataProp"] = false,
+                ["DateTimeOffsetDataProp"] = new DateTimeOffset(2021, 1, 1, 0, 0, 0, TimeSpan.Zero),
+                ["NullableDateTimeOffsetDataProp"] = new DateTimeOffset(2021, 1, 1, 0, 0, 0, TimeSpan.Zero),
+                ["TagListDataProp"] = s_taglist,
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchGenericDataModelMapperTests.cs:68]: nit: For this and the other tests with long list of values would using Theories be better?

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs (hunk 2 above)
Comment (westey-m) [reply dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs]: Updated to null and added a check to make sure it's provided at the end.

File: dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs (hunk 3 above)
Comment (westey-m) [reply dotnet/src/Connectors/Connectors.Memory.AzureAISearch/AzureAISearchGenericDataModelMapper.cs:129]: Updated key to nullable and added a check to make sure it's provided at the end.
That said, there is a lot of validation to make sure that a key is specified.  There may be cases though where there a key defined but it didn't exist in the retrieved record, so adding a check makes sense.

File: dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchGenericDataModelMapperTests.cs (hunk 4 above)
Comment (westey-m) [reply dotnet/src/Connectors/Connectors.AzureAISearch.UnitTests/AzureAISearchGenericDataModelMapperTests.cs:68]: Since the types of fields checked differ and the record has to conform to a certain format, e.g. it always needs a key, theories aren't as great here.  I like theories as well, but I don't think they work well for these tests.

**Review by SergeyMenshykh — APPROVED** [review]

**Review by SergeyMenshykh — APPROVED** [review]
